src/core/valuemapmodelbase.h

```diff
@@ -0,0 +1,84 @@
+/***************************************************************************
+                            valuemapmodelbase.h
+
+                              -------------------
+              begin                : March 2019
+              copyright            : (C) 2019 by Matthias Kuhn
+              email                : matthias@opengis.ch
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef VALUEMAPMODELBASE_H
+#define VALUEMAPMODELBASE_H
+
+#include <QAbstractListModel>
+#include <QVariant>
+
+/**
+ * A model that manages the key/value pairs for a ValueMap widget.
+ */
+class ValueMapModelBase : public QAbstractListModel
+{
+    Q_OBJECT
+
+    /**
+     * A list of QVariantMap, wrapped in a QVariant.
+     *
+     * Like this:
+     *
+     * [{'CH': 'Switzerland'}, {'DE': 'Germany'}, {'FR': 'France'}]
+     */
+    Q_PROPERTY( QVariant valueMap READ map WRITE setMap NOTIFY mapChanged )
+
+  public:
+    /**
+     * Create a new value map model
+     */
+    explicit ValueMapModelBase( QObject *parent = nullptr );
+
+    /**
+     * The map, see the property description
+     */
+
+    QVariant map() const;
+    /**
+     * The map, see the property description
+     */
+    void setMap( const QVariant &map );
```

Comment:
If you remove the Q_PROPERTY, you'll need to slightly tweak the description here.